### PR TITLE
consensus, core, eth/downloader: add 4844 excessDataGas validations

### PR DESCRIPTION
**Pull request by karalabe**
This PR implements validating the excessDataGas field of Cancun headers.

The validations are significantly more complex than previous header validations because the excessDataGas field combines two notions into one: the price for data blobs and the gas used by data blobs. To validate the combo field, we need access to both the parent's excessDataGas as well as the current block's transaction list (blob count). This makes validation more convoluted in both the chain's block processor as well as the downloader, which both needs to juggle two entities (parent header, current block) from now on.

### CODE REVIEW

File: consensus/misc/eip4844.go [hunk 1]
```diff
@@ -27,7 +30,59 @@ var (
...
+	}
+	consumed := big.NewInt(params.BlobTxDataGasPerBlob)
+	consumed.Mul(consumed, big.NewInt(int64(blobs)))
+	excessDataGas.Add(excessDataGas, consumed)
+
+	targetGas := big.NewInt(params.BlobTxTargetDataGasPerBlock)
+	if excessDataGas.Cmp(targetGas) < 0 {
+		return new(big.Int)
+	}
+	return new(big.Int).Sub(excessDataGas, targetGas)
```
Comment (holiman) [inline consensus/misc/eip4844.go:78]:
```suggestion
	return excessDataGas.Sub(excessDataGas, targetGas)
```

File: consensus/misc/eip4844.go [hunk 2]
```diff
@@ -27,7 +30,59 @@ var (
...
+	if parentExcessDataGas != nil {
+		excessDataGas.Set(parentExcessDataGas)
+	}
+	consumed := big.NewInt(params.BlobTxDataGasPerBlob)
+	consumed.Mul(consumed, big.NewInt(int64(blobs)))
+	excessDataGas.Add(excessDataGas, consumed)
+
+	targetGas := big.NewInt(params.BlobTxTargetDataGasPerBlock)
+	if excessDataGas.Cmp(targetGas) < 0 {
+		return new(big.Int)
```
Comment (holiman) [inline consensus/misc/eip4844.go:76]:
```suggestion
		return excessDataGas.SetUint64(0)
```

**Review by holiman — APPROVED** [review]
Looks good, but we need to do a couple of syncs to verify that the changes in `eth/downloader` are ok

File: consensus/misc/eip4844.go [hunk 3]
```diff
@@ -27,7 +30,59 @@ var (
...
+// with no included transactions (e.g. snap sync).
+func VerifyEIP4844Header(parent, header *types.Header) error {
+	// Verify the header is not malformed
+	if header.ExcessDataGas == nil {
+		return errors.New("header is missing excessDataGas")
+	}
+	// Verify the excessDataGas is correct based on the parent header iff the
+	// transaction list is empty. For non-empty blocks, validation needs to be
+	// done later.
+	if header.TxHash == types.EmptyTxsHash {
```
Comment (lightclient) [inline consensus/misc/eip4844.go:49]: It seems pretty rare this code path would actually be triggered, what is benefit of fast failing here?

Comment (karalabe) [reply consensus/misc/eip4844.go:49]: Interestingly, the downloader has an optimisation that it only requests block bodies if they are not empty. If they are empty, then the bodies aren't requested and just pre-filled to zero.

Buuut, 4844's excessDataGas field *needs* the bodies to be verified. Since the downloader just blindly fills teh transaction list as empty in those cases, we have two options:

- Have the excessDataGas already validated like all other header fields (what this check does)
- Introduce header validation pathways into the "fill empty block" mechanism of the downloader

The second thing seemed wonkier as it would need changes that touch a number of unintuitive spots in the code.

But in all honesty, this change is also not that intuitive, it just seemed nicer.

Alternatively we could do something akin to the "gasUsed" which is completely ignored during snap sync and just assumed to be correct whatever it is, but here we not only have "blobsUsed" but also "pricing" so it seemed more relevant to try and validate it.

TL;DR: These are all fallouts of that convoluted field.

File: eth/downloader/queue.go [hunk 4]
```diff
@@ -299,8 +308,8 @@ func (q *queue) Schedule(headers []*types.Header, hashes []common.Hash, from uin
 	for i, header := range headers {
 		// Make sure chain order is honoured and preserved throughout
 		hash := hashes[i]
-		if header.Number == nil || header.Number.Uint64() != from {
-			log.Warn("Header broke chain ordering", "number", header.Number, "hash", hash, "expected", from)
+		if header.Number == nil || header.Number.Uint64() != parent.Number.Uint64()+1 {
+			log.Warn("Header broke chain ordering", "number", header.Number, "hash", hash, "expected", parent.Number)
```
Comment (MariusVanDerWijden) [inline eth/downloader/queue.go:312]:
```suggestion
			log.Warn("Header broke chain ordering", "number", header.Number, "hash", hash, "expected", parent.Number.Uint64()+1)
```

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

File: eth/downloader/queue.go (hunk 4 above)
Comment (karalabe) [reply eth/downloader/queue.go:312]: Good catch

**Comment by karalabe** [comment]
Superseded by https://github.com/ethereum/go-ethereum/pull/27382 according to https://github.com/ethereum/EIPs/pull/7062
